core/libp2p/network/router.hpp

```diff
@@ -6,55 +6,72 @@
+
     virtual ~Router() = default;
 
     /**
-     * @brief Set handler for the {@param protocol}
-     * @param protocol protocol
-     * @param handler function-handler
+     * Set handler for the {@param protocol}; only the perfect handler match
+     * will be invoked
+     * @param protocol to be handled
+     * @param handler to be called, when a new stream for this protocol arrives
      */
-    virtual void setProtocolHandler(
-        const peer::Protocol &protocol,
-        const std::function<connection::Stream::Handler> &handler) = 0;
+    virtual void setProtocolHandler(const peer::Protocol &protocol,
+                                    const ProtoHandler &handler) = 0;
 
     /**
-     * @brief Set handler for the protocol. First, searches all handlers by
-     * given prefix, then executes handler callback for all matches when {@param
-     * predicate} returns true.
+     * Set handler for the <protocol, predicate> pair. First, searches all
+     * handlers by prefix of the given protocol, then executes handler callback
+     * for all matches when {@param predicate} returns true
+     * @param protocol, for which pairs <protocol, handler> are to be retrieved
+     * @param handler to be executed over the protocols, for which the predicate
+     * was evaluated to true
+     * @param predicate to be executed over the found protocols
      */
-    virtual void setProtocolHandlerByPrefix(
-        std::string_view prefix,
-        const std::function<bool(const peer::Protocol &)> &predicate,
-        const std::function<connection::Stream::Handler> &handler) = 0;
+    virtual void setProtocolHandler(const peer::Protocol &protocol,
+                                    const ProtoHandler &handler,
+                                    const ProtoPredicate &predicate) = 0;
 
     /**
-     * @brief Returns list of supported protocols.
+     * Get a list of handled protocols
+     * @return supported protocols; may also include protocol prefixes; if any
+     * set
      */
     virtual std::vector<peer::Protocol> getSupportedProtocols() const = 0;
 
     /**
-     * @brief Removes all associated handlers to given protocol.
-     * @param protocol
+     * Remove handlers, associated with the given protocol prefix
+     * @param protocol prefix, for which the handlers are to be removed
      */
-    virtual void removeProtocolHandler(const peer::Protocol &protocol) = 0;
+    virtual void removeProtocolHandlers(const peer::Protocol &protocol) = 0;
 
     /**
-     * @brief Remove all handlers.
+     * Remove all handlers
      */
     virtual void removeAll() = 0;
 
    protected:
     /**
-     * @brief Execute stored handler for given protocol {@param p}
+     * Execute stored handler for given protocol {@param p}; if several
+     * handlers can be found (for example, if exact protocol match failed, and
+     * prefix+predicate search returned more than one handlers, the random one
```

Comment:
the random one?
why?

why not the first one or last one?

Reply:
Well, actually, in that implementation it's the longest match, so that given "/ping/1.5.5" handler with prefix "/ping/1.5" will have greater priority than the one with prefix "/ping/". 
But, I don't want the users to know about those rules, as the algorithm can change, so there should be no guarantees

Reply:
```
/ping/1
/ping/2
/ping/3
```
with prefix `/ping` in go-libp2p `/ping/3` will be selected (the last one)

Reply:
But is that a specified requirement? Also, we have no way to say, which prefixes were added after which in time, as the library we took compresses them, messing with their order. If we talk about alphabetical order, it can be very consumptive to sort matched handlers each time in order to understand, which one to take. 

So, I am not sure that there is (and should be) some guarantee, which handler is to be called if several have matched.

Reply:
Yeah, okay. Let it be random.